python/gui/MaaS/cbv/DatasetApiView.py

```diff
@@ -0,0 +1,62 @@
+import asyncio
+from django.http import JsonResponse
+from wsgiref.util import FileWrapper
+from django.http.response import StreamingHttpResponse
+from .AbstractDatasetView import AbstractDatasetView
+from .DatasetFileWebsocketFilelike import DatasetFileWebsocketFilelike
+import logging
+logger = logging.getLogger("gui_log")
+
+
+class DatasetApiView(AbstractDatasetView):
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def _get_dataset_content_details(self, dataset_name: str):
+        result = asyncio.get_event_loop().run_until_complete(self.dataset_client.get_dataset_content_details(name=dataset_name))
+        logger.info(result)
+        return JsonResponse({"contents": result}, status=200)
+
+    def _delete_dataset(self, dataset_name: str) -> JsonResponse:
+        result = asyncio.get_event_loop().run_until_complete(self.dataset_client.delete_dataset(name=dataset_name))
+        return JsonResponse({"successful": result}, status=200)
+
+    def _get_datasets_json(self) -> JsonResponse:
+        serial_dataset_map = asyncio.get_event_loop().run_until_complete(self.get_datasets())
+        return JsonResponse({"datasets": serial_dataset_map}, status=200)
+
+    def _get_dataset_json(self, dataset_name: str) -> JsonResponse:
+        serial_dataset = asyncio.get_event_loop().run_until_complete(self.get_dataset(dataset_name=dataset_name))
+        return JsonResponse({"dataset": serial_dataset[dataset_name]}, status=200)
+
+    def _get_download(self, request, *args, **kwargs):
+        dataset_name = request.GET.get("dataset_name", None)
+        item_name = request.GET.get("item_name", None)
+        chunk_size = 8192
+
+        custom_filelike = DatasetFileWebsocketFilelike(self.dataset_client, dataset_name, item_name)
+
+        response = StreamingHttpResponse(
+            FileWrapper(custom_filelike, chunk_size),
+            content_type="application/octet-stream"
+        )
+        response['Content-Length'] = asyncio.get_event_loop().run_until_complete(self.dataset_client.get_item_size(dataset_name, item_name))
+        response['Content-Disposition'] = "attachment; filename=%s" % item_name
+        return response
+
+    def get(self, request, *args, **kwargs):
+        request_type = request.GET.get("request_type", None)
+        if request_type == 'download_file':
```

Comment:
This `if` chain leads me to believe that splitting this out across several smaller views might help.

Reply:
I did originally start to do that, but it occurred to me that this kind of violated the intent of having only a single websocket (at least per authenticated session) open for all client communication.  In fairness, that isn't working properly at the moment, but I think we want to try to have it that way eventually, or else we'd move to REST.  And if we are eventually going for that, I don't think we want too many different views (i.e., because eventually we'll have to consolidate them), although I could be assessing that incorrectly.

Reply:
yeah, if there's a plan to consolidate splitting them out doesn't make a whole lot of sense.